src/SprykerEco/Zed/PunchoutCatalogs/Communication/Form/ConnectionSetupSubForms/PunchoutCatalogConnectionSetupForm.php

```diff
@@ -0,0 +1,352 @@
+            'constraints' => new NotBlank(),
+        ]);
+
+        return $this;
+    }
+
+    /**
+     * @param \Symfony\Component\Form\FormBuilderInterface $builder
+     * @param array $options
+     *
+     * @return $this
+     */
+    protected function addCompanyBusinessUnitField(FormBuilderInterface $builder, array $options)
+    {
+        $builder->add(PunchoutCatalogConnectionSetupTransfer::FK_COMPANY_BUSINESS_UNIT, SelectType::class, array_merge(
+            $this->getCompanyBusinessUnitFieldOptions(),
+            $options
+        ));
+
+        return $this;
+    }
+
+    /**
+     * @param \Symfony\Component\Form\FormBuilderInterface $builder
+     * @param array $options
+     *
+     * @return $this
+     */
+    protected function addCompanyUserField(FormBuilderInterface $builder, array $options)
+    {
+        $builder->add(
+            PunchoutCatalogConnectionSetupTransfer::FK_COMPANY_USER,
+            SelectType::class,
+            array_merge($this->getCompanyUserFieldOptions(), $options)
+        );
+
+        return $this;
+    }
+
+    /**
+     * @param \Symfony\Component\Form\FormBuilderInterface $builder
+     *
+     * @return $this
+     */
+    protected function addCompanyBusinessUnitFieldListeners(FormBuilderInterface $builder)
+    {
+        $formModificationCallback = function (FormEvent $event) {
+            $form = $event->getForm();
+            $loginMode = $event->getData()[PunchoutCatalogConnectionSetupTransfer::LOGIN_MODE] ?? null;
+
+            if ($loginMode !== static::LOGIN_MODE_DYNAMIC_USER) {
+                $this->disableCompanyBusinessUnitField($form);
+
+                return;
+            }
+
+            $idParentCompanyBusinessUnit = $form
+                ->getParent()
+                ->getParent()
+                ->get(PunchoutCatalogConnectionTransfer::FK_COMPANY_BUSINESS_UNIT)
+                ->getData();
+
+            if (!$idParentCompanyBusinessUnit) {
+                return;
+            }
+
+            $this->updateCompanyBusinessUnitFieldChoices($form, $idParentCompanyBusinessUnit);
+        };
+
+        $builder->addEventListener(FormEvents::PRE_SUBMIT, $formModificationCallback);
+        $builder->addEventListener(FormEvents::PRE_SET_DATA, $formModificationCallback);
+
+        return $this;
+    }
+
+    /**
+     * @param \Symfony\Component\Form\FormBuilderInterface $builder
+     *
+     * @return $this
+     */
+    protected function addCompanyUserFieldListeners(FormBuilderInterface $builder)
+    {
+        $formModificationCallback = function (FormEvent $event) {
+            $form = $event->getForm();
+            $loginMode = $event->getData()[PunchoutCatalogConnectionSetupTransfer::LOGIN_MODE] ?? null;
+
+            if ($loginMode !== static::LOGIN_MODE_SINGLE_USER) {
+                $this->disableCompanyUserField($form);
+
+                return;
+            }
+
+            $idParentCompanyBusinessUnit = $form
+                ->getParent()
```

Comment:
getParent()->getParent() already a bad sign. Why do you need this?
Is there any other way of achieving this?

Reply:
Replied here https://github.com/spryker-eco/punchout-catalogs/pull/7#discussion_r294826041